design/implicit-boot-mode.md

```diff
@@ -0,0 +1,197 @@
+  Ironic without relying on the user to set the value directly.
+* Ensure that metal3's behavior is always deterministic.
+
+### Non-Goals
+
+* Address boot mode selection for non-x86 hardware.
+
+## Proposal
+
+For each host, Ironic maintains a set of properties telling it how to
+manage the hardware. One of these is the `boot_mode`. Valid values
+are `bios` and `uefi`. When the `boot_mode` property is
+explicitly set, Ironic will try to ensure that the hardware is
+configured accordingly and it may also choose different paths in the
+provisioning process. In particular, there is a branch in the ISO
+building logic to include different content based on the boot mode.
+
+If the `boot_mode` is not specified, Ironic tries to determine the
+right value by looking at the hardware settings. Some BMC types do not
+support this capability, especially when using the IPMI protocol. When
+Ironic cannot determine the right value for itself, and has not been
+told explicitly what value to use, it falls back to a global default
+specified by a configuration option. Metal3 is not typically deployed
+in a way that allows that option to be changed easily.
+
+We want to ensure that metal3's behavior is always deterministic and
+predictable. We also want to keep the API as minimal as possible, and
+to avoid introducing API fields in a way that makes it likely that a
+user would either need to tell us something we can figure out or that
+they could tell us something incorrect.
+
+Given those goals, it is useful to examine the different situations in
+which different boot modes are appropriate, and how to select the
+right mode at the right time.
+
+The first approach would be to try to always require the same boot
+mode for all hosts. Most modern x86 hosts support UEFI. It is
+therefore useful as a default. However, many contributors have
+somewhat older hardware, which works better with the BIOS boot
+mode. While this hardware is not necessarily our target, it isn't
+practical to abandon support entirely by always requiring
+UEFI. Similarly, there are reasons we cannot always require a host to
+use the BIOS boot mode. UEFI supports desirable features, such as
+secure booting, non-x86 hardware, and PXE on IPv6 networks. For these
+reasons, a single boot mode will not work.
+
+The desired features do give us some hints about how we can determine
+the boot mode implicitly, however. For example, we have no plans to
+support virtual media with IPMI. We can also tell whether the
+provisioning network is using IPv4 or IPv6. Using this information, we
+could build a set of rules to determine the boot mode. However, we
+would have to document those rules, and they may not seem obvious to
+the end-user.
+
+At least some of the features are tied to the protocol used to
+communicate with the BMC. We could therefore link the boot mode to the
+BMC driver, and have a deterministic method for choosing the boot mode
+that is also easy to document.
```

Comment:
This would then mean for supporting BMC protocols other than ```IPMI``` or ```Redfish```, say ```iDRAC-wsman```, it will become *_incumbent_* that this ```boot_mode``` implementation is done. And without it, the implementation would be incomplete?

Reply:
It would be idea, but I don't think required. Documentation can always say "for vendor-z, please ensure hardware has been pre-configured to UEFI mode". Realistically, that will still be required for some firmware versions in BMCs today.

Reply:
The BMC driver here in metal3 will always need the new method to be implemented in order to fulfill the go interface.

Reply:
> 
> 
> The BMC driver here in metal3 will always need the new method to be implemented in order to fulfill the go interface.

That's what I thought. I'll have to take this into consideration then. Let me know whenever the work starts on this, I'd like to contribute to the idrac side of things.